feat: Canvas course export and S3 upload

rachellougee opened the pull request:
### What are the relevant tickets?
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Closes # --->
<!--- Fixes # --->
<!--- N/A --->
https://github.com/mitodl/hq/issues/7459

### Description (What does it do?)
<!--- Describe your changes in detail -->

Implementing a Dagster pipeline to export canvas course content and upload to our designed bucket in S3
- s3://ol-devops-sandbox/ for local testing
- s3://ol-data-lake-landing-zone-production/ for production

For now, we only need to export the course content for https://canvas.mit.edu/courses/7023. We will add more later once Peter give us a list of courses.

### How can this be tested?
<!---
Please describe in detail how your changes have been tested.
Include details of your testing environment, any set-up required
(e.g. data entry required for validation) and the tests you ran to
see how your change affects other areas of the code, etc.
Please also include instructions for how your reviewer can validate your changes.
--->
```
export AWS_ACCESS_KEY_ID=
export AWS_SECRET_ACCESS_KEY=
export GITHUB_TOKEN=
dagster dev -d src/ol_orchestrate/ -f src/ol_orchestrate/definitions/canvas_course_export.py
```
Navigate to http://127.0.0.1:3000/assets/canvas?view=folder and click on "Materialize"
Course ID 7023 should be uploaded to `s3://ol-devops-sandbox/pipeline-storage/canvas/course_export/`

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,98 @@
+            key=AssetKey(["canvas", f"course_{course_id}"]),
+            io_manager_key="s3file_io_manager",
+        )
+        for course_id in course_ids
```
The asset itself is a singleton object. The course IDs are keys in a partition object along with the course asset is segmented. The mental model for an asset is similar to that of a class, where it is the abstract representation of an object. The partitions are instances of the object.

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,98 @@
+
+    extension = "imscc" if export_type == "common_cartridge" else export_type
+
+    for course_id in course_ids:
```
Each execution of this function should only operate against a single course, so this should be something like `course_id = context.partition_key`

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,98 @@
+                time.sleep(30)
+
+        course_export_path = Path(f"{course_id}_course_export.{extension}")
+        response = requests.get(download_url, timeout=20)
```
We've largely been using httpx in this repo, so we should stick with that for consistency.

blarghmatey commented on src/ol_orchestrate/definitions/canvas_course_export.py:
```diff
@@ -0,0 +1,37 @@
+canvas_course_export = Definitions(
+    resources={
+        "io_manager": default_io_manager(DAGSTER_ENV),
+        "s3file_io_manager": S3FileObjectIOManager(
```
This should use the helper method for the IO Manager to make it easier to use across local and deployed environments https://github.com/mitodl/ol-data-platform/blob/integrate-dbt-with-superset/src/ol_orchestrate/lib/dagster_helpers.py#L33

quazi-h reviewed (commented):
I was able to get the pipeline deployed and was able to start an asset materialization job for a canvas course export but the job eventually failed with the following error:
```botocore.exceptions.ReadTimeoutError: Read timeout on endpoint URL: "https://ol-devops-sandbox.s3.amazonaws.com/pipeline-storage/canvas/course_export/7023/ad9cf912abadef30deb0165201951c80e9015f3fbeb6b5fc5eeff12cb032b1bc.imscc?partNumber=1&uploadId=B4O8y.S3xBIu6G3YbKBsmuAkSSTd7tJVi2BYDtTQFnuClIs3gu8OQb9ufOAf6L9GfdAzGeYGUiYll6shSJPsvS3WeAkFasWzZ8BFYbuZDkGW.OdrRrl_EikjTRiHaR06"```
I did try creating a new set of IAM credentials to test with but got the same result, I am not sure why it would time out because Rachel confirmed the jobs ran and completed successfully for her this morning. I wonder if I am setting up the access key incorrectly, I selected the "Local Code" option ("Enable application code in a local development environment to access your AWS account") but maybe it requires the "Third-party service" or "Application running outside AWS" option?

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,94 @@
+)
+
+# predefined course IDs to export
+course_partitions = StaticPartitionsDefinition(["7023"])
```
```suggestion
canvas_course_ids = StaticPartitionsDefinition(["7023"])
```

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,94 @@
+    group_name="canvas",
+    required_resource_keys={"canvas_api"},
+    io_manager_key="s3file_io_manager",
+    partitions_def=course_partitions,
```
```suggestion
    partitions_def=canvas_course_ids,
```

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,94 @@
+    course_id = int(context.partition_key)
+    export_date = datetime.now(tz=UTC).strftime("%Y%m%d")
+    # only export common cartridge for now
+    export_type = "common_cartridge"
```
[nit] this might be good to create as a constant

blarghmatey commented on src/ol_orchestrate/resources/canvas_api.py:
```diff
@@ -0,0 +1,135 @@
+        response.raise_for_status()
+        return response.json()
+
+    def download_course_export(self, url: str, output_path: Path) -> None:
```
[nit] To make it a bit more of a ["fluent API"](https://en.wikipedia.org/wiki/Fluent_interface) it might be useful to return the Path object of the written file.

blarghmatey commented on src/ol_orchestrate/assets/canvas.py:
```diff
@@ -0,0 +1,94 @@
+    )
+
+    with course_export_path.open("rb") as f:
+        data_version = hashlib.file_digest(f, "sha256").hexdigest()
```
To get a proper data version, we would want to do a recursive hash of the contents, excluding the metadata.xml file to avoid erroneous hash changes due to the export timestamp. Something like:

```python
import hashlib
import zipfile
import os
hasher = hashlib.new("sha256")

with zipfile.ZipFile(course_export_path, 'r') as zip_file:
    for item in zip_file.infolist():
        hasher.update(item.filename.encode('utf-8'))
        if not item.is_dir() and not item.filename == "metadata.xml":
            with zip_file.open(item.filename) as file_content:
                while True:
                    chunk = file_content.read(4096)
                    if not chunk:
                        break
                    hasher.update(chunk)

data_version = hasher.hexdigest()
```

It's probably worth building this into a lib module since we'll likely do this for other types of data assets.

rachellougee commented:
@blarghmatey Thanks. I think I’ve addressed all your feedback and ran some tests — the data version seems to be working. Can you please take another look?
